File: project-01-rbac-system/code-java/src/main/java/com/example/demo/service/AuthService.java
package com.example.demo.service;

import org.springframework.stereotype.Service;

import com.example.demo.dto.UserProfileDto;
import com.example.demo.repository.PermissionRepository;
import com.example.demo.repository.UserQueryRepository;

@Service
public class AuthService {

    private final UserQueryRepository userQueryRepository;
    private final PermissionRepository permissionRepository;

    public AuthService(UserQueryRepository userQueryRepository, PermissionRepository permissionRepository) {
        this.userQueryRepository = userQueryRepository;
        this.permissionRepository = permissionRepository;
    }

    public UserProfileDto getUserProfile(String username) {
        return userQueryRepository.getUserProfileWithPermissions(username);
    }

    public boolean checkPermission(Long userId, String resource, String action) {
        return permissionRepository.hasPermission(userId, resource, action);
    }
}

File: project-01-rbac-system/code-kotlin/src/main/kotlin/com/example/demo/service/AuthService.kt
package com.example.demo.service

import com.example.demo.dto.UserProfileDto
import com.example.demo.repository.PermissionRepository
import com.example.demo.repository.UserQueryRepository
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val userQueryRepository: UserQueryRepository,
    private val permissionRepository: PermissionRepository
) {

    fun getUserProfile(username: String): UserProfileDto? {
        return userQueryRepository.getUserProfileWithPermissions(username)
    }

    fun checkPermission(userId: Long, resource: String, action: String): Boolean {
        return permissionRepository.hasPermission(userId, resource, action)
    }
}
